app/controllers/one_login_controller.rb

```diff
@@ -0,0 +1,85 @@
+class OneLoginController < ApplicationController
+  before_action :redirect_to_candidate_sign_in_unless_one_login_enabled
+
+  def callback
+    auth = request.env['omniauth.auth']
+    session[:one_login_id_token] = auth&.credentials&.id_token
+    candidate = OneLoginUser.authenticate(auth)
+
+    sign_in_candidate(candidate)
+
+    redirect_to candidate_interface_interstitial_path
+  rescue OneLoginUser::Error => e
+    session[:one_login_error] = e.message
+    redirect_to auth_one_login_sign_out_path
+  end
+
+  def bypass_callback
+    one_login_user_bypass = OneLoginUserBypass.new(
+      token: request.env['omniauth.auth']&.uid,
+    )
+    candidate = one_login_user_bypass.authenticate
+
+    if one_login_user_bypass.valid? && candidate.present?
+      sign_in_candidate(candidate)
+
+      redirect_to candidate_interface_interstitial_path
+    else
+      flash[:warning] = one_login_user_bypass.errors.full_messages.join('\n')
+      redirect_to candidate_interface_create_account_or_sign_in_path
+    end
+  end
+
+  def sign_out
+    id_token = session[:one_login_id_token]
+    one_login_error = session[:one_login_error]
+    reset_session
+
+    session[:one_login_error] = one_login_error
+    if OneLogin.bypass?
+      redirect_to candidate_interface_create_account_or_sign_in_path
+    else
+      # Go back to one login to sign out the user on their end as well
+      redirect_to logout_one_login(id_token), allow_other_host: true
+    end
+  end
+
+  def sign_out_complete
+    if session[:one_login_error].present?
+      Sentry.capture_message(session[:one_login_error], level: :error)
+      redirect_to internal_server_error_path
+    else
+      redirect_to candidate_interface_create_account_or_sign_in_path
+    end
+  end
+
+  def failure
+    session[:one_login_error] = "One login failure with #{params[:message]} " \
+                                "for one_login_id_token: #{session[:one_login_id_token]}"
```

Comment:
Leaving a note here to remind us to remove this from the logs

Reply:
The one_login_id_token is passed to one login to tell it where to come back after sign-out https://github.com/DFE-Digital/apply-for-teacher-training/blob/4a0da8a64dcd964791f3dd944b4ba94b4f033c99/app/controllers/one_login_controller.rb/#L43